flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,4750 @@
+
+            sage: H.vertical(0).left_half_space()
+            {x ≤ 0}
+
+        """
+        return self.parent().half_space(self._a, self._b, self._c, model="klein")
+
+    def right_half_space(self):
+        # TODO: Should this be in the oriented class? Should there be an equivalent in the unoriented class?
+        r"""
+        Return the closed half space to the right of this (oriented) geodesic.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+            sage: H = HyperbolicPlane(AA)
+
+            sage: H.vertical(0).right_half_space()
+            {x ≥ 0}
+
+        """
+        return (-self).left_half_space()
+
+    def _configuration(self, other):
+        # TODO: Should this be in the oriented class? Should there be an equivalent in the unoriented class?
+        r"""
+        Return a classification of the angle between this
+        geodesic and ``other`` in the Klein model.
+
+        """
+        # TODO: Can we make this public somehow?
+        intersection = self._intersection(other)
+
+        if intersection is None:
+            orientation = (self._b * other._b + self._c * other._c).sign()
+
+            assert orientation != 0
+
+            if self == other:
+                assert orientation > 0
+                return "equal"
+
+            if self == -other:
+                assert orientation < 0
+                return "negative"
+
+            if orientation > 0:
+                return "parallel"
+
+            return "anti-parallel"
+
+        tangent = (self._c, -self._b)
+        orientation = (-tangent[0] * other._b - tangent[1] * other._c).sign()
+
+        assert orientation != 0
+
+        # TODO: Is convex/concave the right term?
+        if orientation > 0:
+            return "convex"
+
+        # TODO: Use the bounding-box trick to get mostly rid of concave cases.
+        return "concave"
+
+    def _intersection(self, other):
+        # TODO: Should this be in the oriented class? Should there be an equivalent in the unoriented class?
+        r"""
+        Return the intersection of this geodesic and ``other`` in the Klein
+        model or in the Euclidean plane if the intersection point is ultra
+        ideal, i.e., not in the unit disk.
+
+        Returns ``None`` if the lines do not intersect in a point.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+            sage: H = HyperbolicPlane(AA)
+
+        ::
+
+            sage: A = -H.vertical(0)
+            sage: B = H.vertical(-1)
+            sage: C = H.vertical(0)
+            sage: A._intersection(B)
+            ∞
+            sage: A._intersection(C)
+            sage: B._intersection(A)
+            ∞
+            sage: B._intersection(C)
+            ∞
+            sage: C._intersection(A)
+            sage: C._intersection(B)
+            ∞
+
+        """
+        if not self.parent().base_ring().is_exact():
```

Comment:
Would it make sense to allow this over non-exact rings? I have polygons over floating point numbers given by an ordered list of half spaces (all of them matter). But plotting just fails
```
from flatsurf.geometry.hyperbolic import HyperbolicPlane
H = HyperbolicPlane(RDF)
g1 = H.geodesic(0,2)
g2 = H.geodesic(1,oo)
h1 = g1.right_half_space() 
h2 = g2.right_half_space()
poly = H.polygon([h1,h2], check=False, assume_sorted=True, assume_minimal=True)
poly.plot()
```

Reply:
I pushed a (hacky draft) version of the predicates idea we discussed. This example plots now.